### PR TITLE
Port DllImportGenerator to dotnet/runtime

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Merge DllImportGenerator into dotnet/runtime with some history re-writing (similar to repo consolidation) to put files into expected locations.

https://github.com/dotnet/runtime/commit/fc18f4de2e701cba67180f1283030854f9bbc0a3 is the commit that merges feature/DllImportGenerator into dotnet/runtime. The commits after that hook up the generator project and the tests to the repo infrastructure and update the solution files in the repo.

This PR should be merged with "Create a Merge Commit".

### CODE REVIEW

**Comment by eerhardt** [comment]
Hmm, why aren't the CI legs running?

**Comment by jkoritzinsky** [comment]
This PR is into the feature branch still, which the CI legs aren't configured to run against. When we PR this feature branch against `main`, CI will run against it.

File: eng/generators.targets [hunk 1]
```diff
@@ -1,45 +1,66 @@
-<Project InitialTargets="ConfigureGenerators">
+<Project>
+
+  <PropertyGroup>
```
Comment (AaronRobinsonMSFT) [inline eng/generators.targets:3]: Has there been an official discussion about this model for enabling generators? This file is named in a generic fashion so I assume the idea is other future generators should be included here. Do we need doc or a sign off for this pattern?

**Review by AaronRobinsonMSFT — APPROVED** [review]
I am signed off but I would like @elinor-fung and perhaps another infrastructure stakeholder to sign off too.

File: eng/generators.targets (hunk 1 above)
Comment (jkoritzinsky) [reply eng/generators.targets:3]: There has not, but this is the first generator that we're enabling across the repo, so we can define the pattern.

I came up with this mechanism after chatting with @ViktorHofer offline to make sure that it is static-graph-compatible (the previous model was not compatible with the NuGet Restore model we use to keep build perf reasonable).

Comment (AaronRobinsonMSFT) [reply eng/generators.targets:3]: @ViktorHofer insight was what I was alluding to. If this pattern was discussed with him I am good to go - :shipit:

File: src/libraries/System.Runtime.InteropServices/tests/Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
+    <IsShipping>false</IsShipping>
+    <IsPackable>false</IsPackable>
+    <!-- We manually enable DllImportGenerator for projects in this folder as part of testing. -->
+    <EnableDllImportGenerator>false</EnableDllImportGenerator>
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/Directory.Build.props:7]: Why do we need to explicitly set these? I thought these would the default for test projects under tests/?

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/Ancillary.Interop.csproj [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <AssemblyName>Microsoft.Interop.Ancillary</AssemblyName>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <RootNamespace>System.Runtime.InteropServices</RootNamespace>
+    <Nullable>enable</Nullable>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Description>APIs required for usage of the DllImportGenerator and related tools.</Description>
+    <DefineConstants>$(DefineConstants);DLLIMPORT_GENERATOR_TESTING</DefineConstants>
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/Ancillary.Interop.csproj:10]: I'll change it to `DLLIMPORT_GENERATOR_TEST` to match `WINHTTPHANDLER_TEST`.

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**